### PR TITLE
[optimization]RestProtocol opt, share client pool among services and avoid potential memory leak.

### PR DESCRIPTION
**Pull request by chickenlj**
An improved version of https://github.com/apache/dubbo/pull/10023

related issue https://github.com/apache/dubbo/issues/9629

### CODE REVIEW

File: dubbo-common/src/main/java/org/apache/dubbo/common/reference/ReferenceCountedResource.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+import org.apache.dubbo.common.logger.Logger;
+import org.apache.dubbo.common.logger.LoggerFactory;
+
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+
+/**
+ * inspired by Netty
+ */
+public abstract class ReferenceCountedResource implements AutoCloseable {
```
Comment (guohao) [inline dubbo-common/src/main/java/org/apache/dubbo/common/reference/ReferenceCountedResource.java:27]: Why not just use netty's `io.netty.util.ReferenceCounted`?

Comment (chickenlj) [reply dubbo-common/src/main/java/org/apache/dubbo/common/reference/ReferenceCountedResource.java:27]: I think what we need is just a common and simplified version for components across the framework to use, so better not rely on Netty.

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/ReferenceCountedClient.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    public boolean isDestroyed() {
+        return resteasyClient.isClosed();
+    }
+
+    @Override
+    protected void destroy() {
+        try {
+            resteasyClient.close();
```
Comment (CrazyHZM) [inline dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/ReferenceCountedClient.java:45]: Add some log output, others look good to me.

**Review by CrazyHZM — APPROVED** [review]

**Review by AlbumenJ — APPROVED** [review]

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java [hunk 3]
```diff
@@ -136,65 +133,77 @@ protected <T> Runnable doExport(T impl, Class<T> type, URL url) throws RpcExcept
 
     @Override
     protected <T> T doRefer(Class<T> serviceType, URL url) throws RpcException {
+        ReferenceCountedClient referenceCountedClient = clients.computeIfAbsent(url.getAddress(), _key -> {
```
Comment (JasonKai9527) [inline dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/RestProtocol.java:136]: if the ip:port is the same,but httpclient config properties are different, how to fix th is situation?

**Pull request merged**